### PR TITLE
gh-118168: Fix Unpack interaction with builtin aliases

### PR DESCRIPTION
**Pull request by JelleZijlstra**
* Issue: gh-118168

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -1786,8 +1786,9 @@ def __typing_unpacked_tuple_args__(self):
         assert self.__origin__ is Unpack
         assert len(self.__args__) == 1
         arg, = self.__args__
-        if isinstance(arg, _GenericAlias):
-            assert arg.__origin__ is tuple
+        if isinstance(arg, (_GenericAlias, types.GenericAlias)):
+            if arg.__origin__ is not tuple:
+                raise TypeError("Unpack[...] must be used with a tuple type")
```
Comment (AlexWaygood) [inline Lib/typing.py:1791]: It feels weirdly inconsistent that we allow nonsensical things with the new syntax, but not with the "backwards-compatible syntax":

```pytb
>>> from typing import TypeVarTuple, Unpack
>>> class Foo[*Ts]: ...
...
>>> Ts = TypeVarTuple("Ts")
>>> Bar = Foo[int, *Ts]
>>> Bar[*list[str]]
__main__.Foo[int, *list[str]]
>>> Bar[Unpack[list[str]]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    Bar[Unpack[list[str]]]
    ~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/cpython/Lib/typing.py", line 394, in inner
    return func(*args, **kwds)
           ~~~~^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/cpython/Lib/typing.py", line 1394, in __getitem__
    args = _unpack_args(args)
           ~~~~~~~~~~~~^^^^^^
  File "/Users/alexw/dev/cpython/Lib/typing.py", line 306, in _unpack_args
    subargs = getattr(arg, '__typing_unpacked_tuple_args__', None)
              ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alexw/dev/cpython/Lib/typing.py", line 1791, in __typing_unpacked_tuple_args__
    raise TypeError("Unpack[...] must be used with a tuple type")
TypeError: Unpack[...] must be used with a tuple type
```

But I guess that's a preexisting problem, not new to this PR. We previously discussed the issue of "all generic aliases being unpackable" in https://github.com/python/cpython/issues/103450.

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -978,6 +978,37 @@ def foo(**kwargs: Unpack[Movie]): ...
...
+
+    def test_unpack_wrong_type(self):
+        Ts = TypeVarTuple("Ts")
+        class Gen[*Ts]: ...
+        PartGen = Gen[int, *Ts]
+
+        with self.assertRaises(TypeError):
+            PartGen[Unpack[list[int]]]
+        with self.assertRaises(TypeError):
+            PartGen[Unpack[List[int]]]
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
        bad_unpack_param = re.escape("Unpack[...] must be used with a tuple type")
        with self.assertRaisesRegex(TypeError, bad_unpack_param):
            PartGen[Unpack[list[int]]]
        with self.assertRaisesRegex(TypeError, bad_unpack_param):
            PartGen[Unpack[List[int]]]
```

**Review by AlexWaygood — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
@AlexWaygood thanks for the review! Do you think this is safe to backport? (I'm leaning yes.)

**Comment by AlexWaygood** [comment]
> @AlexWaygood thanks for the review! Do you think this is safe to backport? (I'm leaning yes.)

yeah, I think should be fine

**Pull request merged**